### PR TITLE
Improve changelog generation

### PR DESCRIPTION
**Pull request by oandregal**
Part of https://github.com/WordPress/gutenberg/issues/27448

This PR tries to improve the changelog generated for each release, so it needs less manual intervention. These are the sections the changelog would produce with this PR's changes:

Section | Labels
---------- | ---------
Features | [Type] Feature
Enhancements | [Type] Enhancement
New APIs | [Type] New API
Bug Fixes | [Type] Bug, [Type] Regression, also tries to infer from the title when it contains variations of "fix" or "bug"
Performance | [Type] Performance
Experiments | [Type] Experimental, [Block] Navigation, [Block] Query, [Block] Post Comments Count, [Block] Post Comments Form, [Block] Post Comments, [Block] Post Featured Image, [Block] Post Hierarchical Terms, [Block] Post Title, [Block] Site Logo, [Feature] Full Site Editing, Global Styles, [Feature] Navigation Screen, [Feature] Widgets Screen
Documentation | [Type] Documentation
Code Quality | [Type] Code Quality
Tools | [Package] Dependency Extraction Webpack Plugin, [Package] Jest Puppeteer aXe, [Package] E2E Tests, [Package] E2E Test Utils, [Package] Env, [Package] ESLint plugin, [Package] stylelint config, [Package] Project management automation, '[Type] Project Management, [Package] Scripts, [Type] Build Tooling, Automated Testing
Security | [Type] Security
Breaking Change | [Type] Breaking Change
Various | PRs that haven't been classified by any other tag or title inference.

## What I've done

- I've taken a look at [the release posts on the first page of make core](https://make.wordpress.org/core/tag/gutenberg-new/).
- Reviewed the sections to understand whether the PRs were there because they had the proper labels or not (hence required manual curation). I wanted to grow an intuition on where things stood.
- My impression is that:
  - Some sections were working fine (PRs had labels related to the section most of the time): bug fixes, documentation, code quality, etc.
  - Build tooling/tools section: this could work better if we used more labels to classify them (see table above).
  - Experiments section: this could work better if we organized the PRs according to phase 2 milestones ([see](https://github.com/WordPress/gutenberg/issues/27448)). See the new sections "FSE: ..." and "Screen: ..." (note that while the changelog generation creates new sections for them, they could be part of the "experiments" section in the release post easily see "Next Steps").
  - Mobile section: I realized we remove all the PRs related to mobile, so these were grouped under their own section to inspect and remove them easily.
  - Some other sections require more thought and can be iterated later: features, enhancements, new APIs, etc.

## How to test

- Run `./bin/commander.js changelog` and review what the 9.7 would look like.
- Run `./bin/commander.js changelog -m "Gutenberg 9.6"` and compare with the [release post for 9.6](https://make.wordpress.org/core/2020/12/23/whats-new-in-gutenberg-23-december/). Does it feel like a good match?
- Run `./bin/commander.js changelog -m "Gutenberg 9.5"` and compare with the [release post for 9.5](https://make.wordpress.org/core/2020/12/02/whats-new-in-gutenberg-2-december/). Does it feel like a good match?
- Run `./bin/commander.js changelog -m "Gutenberg 9.4"` and compare with the [release post for 9.4](https://make.wordpress.org/core/2020/11/19/whats-new-in-gutenberg-18-november-2/). Does it feel like a good match?

## Next steps

The next step I'd like to address is the ability to group PRs within the same section. For example, group all PRs that affect the same block in the "bug fixes" section, or perhaps convert "FSE: ...", "Screen: ..." sections to groups within the "Experiments" category, etc.

I'd like to address this separately but the changes here should already be useful in reducing the amount of time involved in preparing the release post.

### CODE REVIEW

**Comment by youknowriad** [comment]
> Accessibility

I don't think we need an accessibility category in the  changelog but we can use this to group items inside "bugs" or "enhancements".

> Experimental

Sometimes something is both experimental and a bug fix (or enhancement), I believe we should give priority to "Experimental" in this case. Does the tool do that? 

> FSE: * sections

These should be subsections of the "Experimental" section, is this the case? Same for screen navigation and widgets

> Breaking change

This  doesn't mean anything for me in terms of changelog and I believe we should remove it.

> Mobile

These should be removed from the changelog entirely, mobile apps have their own process for releases.

Thanks for your work here.

**Comment by nosolosw** [comment]
**Accessibility, Experimental, and Breaking Change now work as you suggested.** In case you want to take a look and suggest other changes, this is how the categorization mechanism works: if a PR has many labels, the one that has a lower index in the `LABEL_MAPPING` object will be chosen ― it fallbacks to "Various" if the PR doesn't have any label present in `LABEL_MAPPING`.

**"FSE: ..." and "Screen: ..."**. With the current setup, it's straightforward to make these sub-headings within the experimental section (add one `#`) or make them a group (remove the `###`) within the section. Would it be ok to land this as it is and iterate to add grouping in a subsequent PR?

**Mobile**. Now that they are grouped within a category, deleting them is straightforward. I didn't remove it in case the release manager wants to double check them.

Thoughts? Good to go and iterate?

**Comment by youknowriad** [comment]
> Would it be ok to land this as it is and iterate to add grouping in a subsequent PR?

I think in the previous release changelogs, we did group these in experiments but manually. It would be good if we can stay consistent.

> Mobile

I believe the previous version of the changelog command did remove these.

--

I'll give this a try in a few minutes.

**Comment by youknowriad** [comment]
Just gave this a test and here's what I don't like:

 - There's too many top level sections, I definitely think we should move the "FSE*" and screens under experiment.
 - The removal of "accessibility" means these items are dispatched in the other sections, what would be cool is to group them (sub lists) under the parent sections (bug, enhancement).
 - I like the renaming of "Tools" instead of "Build Tools".
 - Mobile I still think we should remove as it doesn't bring any value to plugin users (these changelogs are meant for plugin releases)

It seems there's an opportunity to do better in terms of grouping based on "blocks" and  things like that but that's something for later.

**Comment by nosolosw** [comment]
@youknowriad addressed your suggestions, so as it is, it should reduce the amount of time required to re-categorize issues.

File: bin/plugin/commands/changelog.js [hunk 1]
```diff
@@ -47,20 +47,51 @@ const manifest = require( '../../../package.json' );
...
-	Task: 'Various',
+	'[Block] Navigation': 'Experiments',
+	'[Block] Query': 'Experiments',
+	'[Block] Post Comments Count': 'Experiments',
+	'[Block] Post Comments Form': 'Experiments',
+	'[Block] Post Comments': 'Experiments',
+	'[Block] Post Featured Image': 'Experiments',
+	'[Block] Post Hierarchical Terms': 'Experiments',
+	'[Block] Post Title': 'Experiments',
+	'[Block] Site Logo': 'Experiments',
```
Comment (youknowriad) [inline bin/plugin/commands/changelog.js:68]: Just tested the PR and it works great. as a follow-up we might want to group more things automatically for instance, create a sublist for all FSE related block PRs 

``
### Experiments:

 - Full Site Editing Blocks:
    - entry 1
    - entry 2
```

There are more things that could use this "sub grouping".

**Review by youknowriad — APPROVED** [review]

**Pull request merged**